spec/firebase_app_distribution_action_spec.rb

```diff
@@ -1,141 +1,131 @@
+  let(:fake_connection) { double("Connection") }
+  let(:fake_binary) { double("Binary") }
 
-  before do
-    allow(Tempfile).to receive(:new).and_return(fake_file)
-    allow(fake_file).to receive(:unlink)
-    allow(fake_file).to receive(:path).and_return("/tmp/string")
+  before(:each) do
+    allow(Fastlane::Actions::FirebaseAppDistributionAction).to receive(:connection).and_return(fake_connection)
+    allow(fake_binary).to receive(:read).and_return("Hello World")
+    allow(File).to receive(:open).and_return(fake_binary)
   end
 
-  describe '#run' do
-    it 'shells out to firebase' do
-      expect(Fastlane::Actions).to receive(:sh_control_output).with("/tmp/fake-firebase-cli appdistribution:distribute /tmp/FakeApp.ipa --app abc:123 --testers-file /tmp/testers.txt --release-notes-file /tmp/release_notes.txt --token fake-token --debug", { print_command: false, print_command_output: true })
-      params = {
-        app: "abc:123",
-        ipa_path: "/tmp/FakeApp.ipa",
-        firebase_cli_path: "/tmp/fake-firebase-cli",
-        testers_file: "/tmp/testers.txt",
-        release_notes_file: "/tmp/release_notes.txt",
-        firebase_cli_token: "fake-token",
-        debug: true
-      }
-      Fastlane::Actions::FirebaseAppDistributionAction.run(params)
-    end
-
-    it 'removes trailing newlines from firebase_cli_path' do
-      expect(Fastlane::Actions).to receive(:sh_control_output).with("/tmp/fake-firebase-cli appdistribution:distribute /tmp/FakeApp.ipa --app abc:123 --testers-file /tmp/testers.txt --release-notes-file /tmp/release_notes.txt --token fake-token --debug", { print_command: false, print_command_output: true })
-      params = {
-          app: "abc:123",
-          ipa_path: "/tmp/FakeApp.ipa",
-          firebase_cli_path: "/tmp/fake-firebase-cli\n",
-          testers_file: "/tmp/testers.txt",
-          release_notes_file: "/tmp/release_notes.txt",
-          firebase_cli_token: "fake-token",
-          debug: true
-      }
-      Fastlane::Actions::FirebaseAppDistributionAction.run(params)
-    end
-  end
-
-  describe "flag helpers" do
-    let(:action) { Fastlane::Actions::FirebaseAppDistributionAction }
-
-    describe "flag_value_if_supplied" do
-      it "returns flag and param value when it exists" do
-        params = {
-            firebase_cli_token: 'fake-token'
-        }
-        expect(action.flag_value_if_supplied('--token', :firebase_cli_token, params)).to eq("--token fake-token")
-      end
-
-      it "returns nil if param does not exist" do
-        params = {}
-        expect(action.flag_value_if_supplied('--token', :firebåse_cli_token, params)).to be_nil
+  describe '#get_upload_token' do
+    context 'when testing with valid parameters' do
+      it 'should make a GET call to the app endpoint and return the upload token' do
+        expect(fake_connection).to receive(:get)
+          .with("/v1alpha/apps/app_id")
+          .and_return(
+            double("Response", status: 200, body: {
```

Comment:
Consider instantiating and returning actual Response class of the type returned by the library you're using instead of returning a test double, which can't be guaranteed to behave like the code under test would.